envoyauth/response.go

```diff
@@ -346,74 +310,64 @@ func (result *EvalResult) GetResponseEnvoyHTTPStatus() (*ext_type_v3.HttpStatus,
 	return status, nil
 }
 
-func transformToHTTPHeaderFormat(input interface{}, result *http.Header) error {
+func makeHeaderValueOption(k, v string) *ext_core_v3.HeaderValueOption {
+	return &ext_core_v3.HeaderValueOption{
+		Header: &ext_core_v3.HeaderValue{
+			Key:   k,
+			Value: v,
+		},
+	}
+}
 
-	takeResponseHeaders := func(headers map[string]interface{}, targetHeaders *http.Header) error {
-		for key, value := range headers {
-			switch values := value.(type) {
-			case string:
-				targetHeaders.Add(key, values)
-			case []string:
-				for _, v := range values {
-					targetHeaders.Add(key, v)
-				}
-			case []interface{}:
-				for _, value := range values {
-					if headerVal, ok := value.(string); ok {
-						targetHeaders.Add(key, headerVal)
-					} else {
-						return fmt.Errorf("invalid value type for header '%s'", key)
-					}
+func makeEnvoyHeaderValueOptionsFromHeadersMap(hvo []*ext_core_v3.HeaderValueOption, headers map[string]any) ([]*ext_core_v3.HeaderValueOption, error) {
+	hvo = slices.Grow(hvo, len(headers))
+	for key, value := range headers {
+		key = textproto.CanonicalMIMEHeaderKey(key)
```

Comment:
If we didn't do this before, it might be worth calling it out on the change log. It's the right thing to do, I suppose, but could potentially break someone's policy if they depended on the headers coming out like they went in.

Reply:
Previously the plugin used `http.Header` where the function `textproto.CanonicalMIMEHeaderKey` has been approached internally.

Reply:
Ah wait, I was misreading this before -- we're normalizing the header keys "on the way out", so when preparing the CheckResponse for Envoy. So, it probably isn't needed -- [Envoy will normalize the keys to all-lowercase anyways](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_conn_man/header_casing), if I'm reading this right.

So the only effect here would be case-related clashes -- if a policy returns headers like `{"foo": "bar", "Foo": "baz"}`, then this line will make the output `{"Foo": "..."}` with a non-deterministic outcome. Before, we'd rely on Envoy to decide, I guess?

In the interest of caution, I'd keep the current behaviour and drop this line. What do you think?

Reply:
I've removed normalization of the key.